### ISSUE TITLE
MMLU-Pro Few-Shot Accuracy: 0.1011–0.1612 (Without lm-eval-harness)

### ISSUE THREAD

**Issue by jasonrichdarmawan** [seal-rg/recurrent-pretraining#26]
Note: I am going to test it with lm-eval-harness this week

Hi,

My goal is to understand how **huginn-0125** works and to confirm whether the method I’m using to extract steering vectors is correct—for example, if performance improves, it would indicate that the extraction method is likely working properly.

Unfortunately, the accuracy without any intervention using the steering vectors remains very low (ranging from 0.1011 to 0.1612). I suspect there may be an issue with how I’ve structured the `system_prompt`, `fewshot_prompts`, and `question_content`.

For your reference:
1. I set the `system_prompt` to: `You are a helpful assistant.`
2. I noticed that you're using `evaluator.simple_evaluate`, where the parameter `fewshot_as_multiturn` defaults to `False`. Accordingly, I placed the `fewshot_prompts` all in a single turn.
3. The `question_content` follows the format: `Q: ...\nA: `, after which the model proceeds to the next turn (Huginn's response).

Here is an example of how the fewshot_prompts are formatted in a single turn:
```
Q: ...
A: ...
```

[The code to generate the queries](https://github.com/jasonrichdarmawan/steering-vectors-on-recurrent-model/blob/1acba3baf51d34dcd1ff967fe8ea81c42f6e1684/utils/prepare_queries.py)

[The code to generate the sentences](https://github.com/jasonrichdarmawan/steering-vectors-on-recurrent-model/blob/1acba3baf51d34dcd1ff967fe8ea81c42f6e1684/utils/generate_sentences.py)

`prepare_queries.py`

```python
def generate_query(
  model_name: str,
  question_content: str,
  tokenizer: PreTrainedTokenizerBase | None = None,
  system_prompt: str | None = None,
  fewshot_prompts: list[str] | None = None,
  with_cot: bool = True
):
  """
  Constructs the final query string for the model based on the provided parameters.

  Args:
    model_name: The identifier for the model architecture, typically obtained from PreTrainedModel.config.model_type. 
      This determines how the query is formatted and which prompt templates or special handling are applied 
      (e.g., chat-style formatting for models starting with "huginn-").
    question_content: The main content of the question to be asked (do NOT add "Q: " prefix; formatting is handled here).
    tokenizer: Optional tokenizer for formatting chat templates.
    system_prompt: Optional system prompt to guide the model's behavior.
    fewshot_prompts: Optional list of few-shot example prompts to include before the main question.
    with_cot: If True, encourages chain-of-thought reasoning by appending "Let's think step by step." to the prompt.

  Returns:
    The fully formatted query string ready for model inference.

  Notes:
    - For models starting with "huginn_", the function constructs a chat-style prompt using a system message and a user message.
    - If few-shot prompts are provided, they are prepended to the user message.
    - If chain-of-thought reasoning is enabled, the question is followed by "A: Let's think step by step." to encourage detailed reasoning.
    - Otherwise, the question is followed by "A: " for a direct answer.
    - The tokenizer's chat template is used to format the final query string.
  """
  query: str = ""

  match model_name:
    case name if name.startswith("huginn_"):
      """
      Reference: 
      [1] https://github.com/seal-rg/recurrent-pretraining/blob/0d9ed974d253e16498edec5c0c0916fdef4eb339/examples/inference_demo.ipynb
      [2] https://github.com/yihuaihong/Linear_Reasoning_Features/blob/f23f2547862a2c1b57f1cfa3c547776cb38f762a/reasoning_representation/Intervention/utils.py
      """
      messages: list[str] = []
      
      if system_prompt == None:
        system_prompt = "You are a helpful assistant."
      
      messages.append({
        "role": "system",
        "content": system_prompt
      })

      user_content = ""

      if fewshot_prompts:
        for fewshot_prompt in fewshot_prompts:
          # Each few-shot prompt should already end with "\n\n"
          user_content += fewshot_prompt
      
      # If few-shot prompts are provided and chain-of-thought (CoT) reasoning is enabled (with_cot=True),
      # append the question followed by "A: Let's think step by step." to encourage step-by-step reasoning.
      # Otherwise, append the question followed by "A: " for a direct answer.
      if fewshot_prompts and with_cot:
        user_content += f"Q: {question_content}\nA: Let's think step by step. "
      else:
        user_content += f"Q: {question_content}\nA: "

      messages.append({
        "role": "user",
        "content": user_content
      })

      query: str = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=True
      )
    case _:
      raise ValueError(f"Unsupported model name: {model_name}")
    
  return query
```

`generate_sentences.py`

```
def generate_sentences(
  model: nn.Module,
  tokenizer: PreTrainedTokenizerBase,
  text: list[str],
  config: GenerationConfig,
):
  match model.config.model_type:
    case name if "huginn_" in name:
      """
      Reference: 
      [1] https://github.com/yihuaihong/Linear_Reasoning_Features/blob/f23f2547862a2c1b57f1cfa3c547776cb38f762a/reasoning_representation/Intervention/utils.py
      [2] https://github.com/seal-rg/recurrent-pretraining/blob/9f84159bc548f4fe75a577d71575c35ef80e1977/examples/inference_demo.ipynb
      """
      input_ids = tokenizer(
        text,
        return_tensors="pt",
        add_special_tokens=False,
        padding="longest",
        return_token_type_ids=False,
      ).input_ids.to("cuda")

      with torch.no_grad():
        outputs = model.generate(
          input_ids,
          config,
          num_steps=32,
          tokenizer=tokenizer
        )

      generated_texts = tokenizer.batch_decode(
        outputs.sequences[:, input_ids.shape[1]:],
        skip_special_tokens=True,
      )
    case _:
      raise ValueError(f"Unsupported model type: {model.config.model_type}")
    
  return generated_texts
```

**Comment by JonasGeiping** [comment]
All things considered, MMLU-pro is a hard tasks for this model size - I am not quite sure what the accuracy of the model is. Can you reproduce a MMLU (normal, 5-shot) accuracy of 38%. This would be a basic test, without chat template, without system prompt - just the few shot prompts as in lm-eval.

Regarding your format, it looks sensible, but why not reproduce the established format (for either MMLU or MMLU-pro), as in https://github.com/EleutherAI/lm-evaluation-harness/blob/9fbe48c230c2649d9430c133290d6882b55105ea/lm_eval/tasks/mmlu_pro/_default_template_yaml#L9 and https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/tasks/mmlu_pro/mmlu_pro_biology.yaml?

**Comment by jasonrichdarmawan** [comment]
> All things considered, MMLU-pro is a hard tasks for this model size - I am not quite sure what the accuracy of the model is. Can you reproduce a MMLU (normal, 5-shot) accuracy of 38%? This would be a basic test, without chat template, without system prompt - just the few shot prompts as in lm-eval.

Hi, I need to compare my results against a previous method, so I’m using two evaluators:
1. The lm-eval-harness library
2. Custom code based on the previous method.

The problem is that the model’s accuracy worsens when it is intervened with a steering vector in my custom evaluation setup.

Note: When using the lm-eval-harness benchmark, the model’s accuracy actually improves slightly with the steering vector, by 1.4% on the MMLU benchmark.

Would you mind to check whether the code below is correct for generating text without a chat template, without a system prompt, but with few-shot prompts?

https://github.com/jasonrichdarmawan/steering-vectors-on-recurrent-model/blob/72a4c3353297dcd74a3b950666f19435ec5a5c3f/utils/generate_sentences.py#L5-L52

```
def generate_sentences_huginn(
  model: nn.Module,
  tokenizer: PreTrainedTokenizerBase,
  text: list[str],
  num_steps: int | None = None
):
  """
  Reference: 
  [1] https://github.com/yihuaihong/Linear_Reasoning_Features/blob/f23f2547862a2c1b57f1cfa3c547776cb38f762a/reasoning_representation/Intervention/utils.py
  [2] https://github.com/seal-rg/recurrent-pretraining/blob/9f84159bc548f4fe75a577d71575c35ef80e1977/examples/inference_demo.ipynb
  [3] https://github.com/yihuaihong/Linear_Reasoning_Features/blob/main/reasoning_representation/Intervention/features_intervention.py
  [4] https://github.com/seal-rg/recurrent-pretraining/blob/0d9ed974d253e16498edec5c0c0916fdef4eb339/evaluate_raven/hf_eval_adaptive_compute.py
  """
  config = GenerationConfig(
    max_new_tokens=200,
    stop_strings=["<|end_text|>", "<|end_turn|>"],
    do_sample=False,
    temperature=None,
    top_p=None,
    min_p=None,
    return_dict_in_generate=False,
    eos_token_id=65505,
    bos_token_id=65504,
    pad_token_id=65509
  )

  input_ids = tokenizer(
    text,
    return_tensors="pt",
    add_special_tokens=False,
    padding="longest",
    return_token_type_ids=False,
  ).input_ids.to(model.device)

  with torch.no_grad():
    outputs = model.generate(
      input_ids,
      config,
      num_steps=num_steps,
      tokenizer=tokenizer,
    )

  generated_texts = tokenizer.batch_decode(
    outputs[:, input_ids.shape[1]:],
    skip_special_tokens=True,
  )

  return generated_texts
```

**Comment by JonasGeiping** [comment]
You have set `add_special_tokens=False`.  Is the `bos` token still correctly added to the beginning of each sequence?

**Comment by jasonrichdarmawan** [comment]
I didn’t prepend the `bos` token to the text argument. I genuinely didn’t realize this mistake. Thank you!

=== out of topic

I was writing the code based on the following example, which uses a chat template.

https://github.com/seal-rg/recurrent-pretraining/blob/9f84159bc548f4fe75a577d71575c35ef80e1977/examples/inference_demo.ipynb

=== out of topic

According to the code below, `bos` token is `<|begin_text|>`

```
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained(LOCAL_PATH)
tokenizer.decode(65504)
```
